aten/src/ATen/StorageImpl.h

```diff
@@ -43,13 +43,14 @@ struct AT_API StorageImpl : public c10::raw_intrusive_ptr_target<StorageImpl> {
   // the real data shouldn't call th::from_type
   template <typename T>
   inline T* data() const {
-    auto scalar_type_T = at::CTypeToScalarType<th::from_type<T>>::to();
-    if (scalar_type_ != scalar_type_T) {
+    auto data_type_T =
+        at::scalarTypeToDataType(at::CTypeToScalarType<th::from_type<T>>::to());
```

Comment:
Hm... when can we get rid of the use of `th::from_type<T>` here?

Reply:
I'm looking into this right now.  It's not trivial because TH actually calls this function with <THHalf>.  We could intermediate this function with a function in TH that we always call, but hopefully we don't have to do that.